src/token.sw

```diff
@@ -25,3 +27,66 @@ pub fn force_transfer(amount: u64, asset_id: ContractId, contract_id: ContractId
         tr r3 r1 r2;
     }
 }
+
+/// Transfer `amount` coins of type `asset_id` to address `recipient`.
+pub fn transfer_to_output(amount: u64, asset_id: ContractId, recipient: Address) {
+    // note: if tx format changes, the magic number "48" must be changed !
+    // TransactionScript outputsCount has a 48 byte(6 words * 8 bytes) offset
+    // Transaction Script: https://github.com/FuelLabs/fuel-specs/blob/master/specs/protocol/tx_format.md#transactionscript
+    // Output types: https://github.com/FuelLabs/fuel-specs/blob/master/specs/protocol/tx_format.md#output
+    const OUTPUT_LENGTH_LOCATION = 48;
+    const OUTPUT_VARIABLE_TYPE = 4;
+
+    // get length of outputs from TransactionScript outputsCount:
+    let length: u8 = asm(outputs_length, outputs_length_ptr: OUTPUT_LENGTH_LOCATION) {
+        lw outputs_length outputs_length_ptr i0;
+        outputs_length: u8
+    };
+    // maintain a manual index as we only have `while` loops in sway atm:
+    let mut index: u8 = 0u8;
+    let mut outputIndex = 0;
+    let mut output_found = false;
+
+    // If an output of type `OutputVariable` is found, check if its `amount` is zero.
+    // As one cannot transfer zero coins to an output without a panic, a variable output with a value of zero is by definition unused.
+    while index < length {
+        let target_output_type_exists = asm(slot: index, type, target: OUTPUT_VARIABLE_TYPE, bytes: 8, res) {
+            xos type slot;
+            meq res type target bytes;
+            res: bool
+        };
```

Comment:
I don't think the above is doing what you want it to.

`xos type slot`  will get the offset to the `slot`th output value and put it in `type`.

`meq res type target bytes`  will compare the two memory regions at `type` and `target` of `bytes` (8) bytes and put it in `res`.  `type` is the offset to the output value, but `target` is 4 -- the value you want to compare against, not an offset.  Since it's a number, and not a struct, array or anything larger than a word, then `target` will be passed to the asm block by value.  I might be wrong, but this is my understanding.

And also according to that spec, the output type is a single byte, so you shouldn't be comparing 8 bytes.

What I think you want is to call `xos` and then just `lb` on what it returns to read the first byte in the output, which is the type.  Then compare it to `OUTPUT_VARIABLE_TYPE` outside the asm block.

Reply:
fixed.